agent/agent.go

```diff
@@ -528,153 +571,117 @@ L:
+		return status, trace.Wrap(err, "failed to record local timeline events")
+	}
+
+	if err := r.notifyMasters(ctx); err != nil {
+		return status, trace.Wrap(err, "failed to notify master nodes of local timeline events")
+	}
+
+	return status, nil
 }
 
 // getLocalStatus obtains local node status.
 func (r *agent) getLocalStatus(ctx context.Context, local serf.Member, respc chan<- *statusResponse) {
-	status := r.collectLocalStatus(ctx, &local)
+	status, err := r.collectLocalStatus(ctx, &local)
 	resp := &statusResponse{
 		NodeStatus: status,
 		member:     local,
+		err:        err,
 	}
 	select {
 	case respc <- resp:
 	case <-r.done:
 	}
 }
 
-// getStatusFrom obtains node status from the node identified by member.
-func (r *agent) getStatusFrom(ctx context.Context, member serf.Member, respc chan<- *statusResponse) {
-	client, err := r.dialRPC(ctx, &member)
-	resp := &statusResponse{member: member}
+// notifyMasters pushes new timeline events to all master nodes in the cluster.
+func (r *agent) notifyMasters(ctx context.Context) error {
+	members, err := r.SerfClient.Members()
 	if err != nil {
-		resp.err = trace.Wrap(err)
-	} else {
-		defer client.Close()
-		var status *pb.NodeStatus
-		status, err = client.LocalStatus(ctx)
-		if err != nil {
-			resp.err = trace.Wrap(err)
-		} else {
-			resp.NodeStatus = status
-		}
-	}
-	select {
-	case respc <- resp:
-	case <-r.done:
-	}
-}
-
-// timelineUpdateLoop periodically updates the status timeline.
-func (r *agent) timelineUpdateLoop(ctx context.Context) {
-	ticker := r.Clock.NewTicker(statusUpdateTimeout)
-	defer ticker.Stop()
-
-	for range ticker.Chan() {
-		if utils.IsContextDone(ctx) {
-			log.Info("Timeline update loop is stopping.")
-			return
-		}
-		if err := r.updateTimeline(ctx); err != nil {
-			log.WithError(err).Error("Failed to update timeline.")
-		}
+		return trace.Wrap(err)
 	}
-}
 
-// updateTimeline records any status changes to the timeline.
-func (r *agent) updateTimeline(ctx context.Context) error {
-	if err := r.updateLocalChanges(ctx); err != nil {
-		return trace.Wrap(err, "error updating local timeline changes")
-	}
-	r.memberIndex++
-	member, err := r.selectMember(int(r.memberIndex))
+	events, err := r.LocalTimeline.GetEvents(ctx, nil)
 	if err != nil {
-		return nil
-	}
-	if err := r.updateRemoteChanges(ctx, member); err != nil {
-		return trace.Wrap(err, "error updating remote timeline changes")
+		return trace.Wrap(err)
 	}
-	return nil
-}
 
-// updateLocalChanges updates the timeline with any new status changes found on
-// the local node.
-func (r *agent) updateLocalChanges(ctx context.Context) error {
-	ctxUpdate, cancel := context.WithTimeout(ctx, updateTimelineTimeout)
-	defer cancel()
-
-	if err := r.Timeline.RecordStatus(ctxUpdate, r.LocalStatus()); err != nil {
-		return trace.Wrap(err)
+	// TODO: async
+	for _, member := range members {
+		if role, ok := member.Tags["role"]; ok && Role(role) == RoleMaster {
```

Comment:
I would either consider moving this condition into a helper method, or even adding a convenient helper `Masters()` method on the serf client.

Also, do we need to only consider "active" members here?